src/mono/mono/component/mini-wasi-debugger.c

```diff
@@ -0,0 +1,172 @@
+#include <glib.h>
+#include <mono/mini/debugger-agent-external.h>
+#include <mono/component/debugger-engine.h>
+#include <mono/metadata/components.h>
+
+#ifdef HOST_WASI
+static int conn_fd;
+static int log_level = 1;
+static int retry_receive_message = 100;
+static int connection_wait = 1000;
+
+__attribute__((import_module("wasi_snapshot_preview1")))
+__attribute__((import_name("sock_accept")))
+int sock_accept(int fd, int fdflags, int* result_ptr);
+
+static long long
+timeInMilliseconds ()
+{
+	struct timeval tv;
+	gettimeofday(&tv,NULL);
+	return (((long long)tv.tv_sec)*1000)+(tv.tv_usec/1000);
+}
+
+static int
+wasi_transport_recv (void *buf, int len)
+{
+	int res;
+	int total = 0;
+	int fd = conn_fd;
+	int again = 0;
+
+	do {
+		res = read (fd, (char *) buf + total, len - total);
+		if (res > 0)
+			total += res;
+		again++;
+		if ((res > 0 && total < len) || (res == -1 && again < retry_receive_message)) {
+			// Wasmtime on Windows doesn't seem to be able to sleep for short periods like 1ms so we'll have to spinlock
+			long long start = timeInMilliseconds ();
+			while (timeInMilliseconds () < start + (connection_wait/1000));
+		} else {
+			break;
+		}
+	} while (true);
+	return total;
+}
+
+static gboolean
+wasi_transport_send (void *data, int len)
+{
+	int res;
+	do {
+		res = write (conn_fd, (const char*)data, len);
+	} while (res == -1);
+
+	return res == len;
+}
+
+/*
+ * wasi_transport_connect:
+ *
+ *   listen on the socket that is already opened by wasm runtime
+ */
+static void
+wasi_transport_connect (const char *socket_fd)
+{
+	PRINT_DEBUG_MSG (1, "wasi_transport_connect.\n", conn_fd);
+	bool handshake_ok = FALSE;
+	
+	while (!handshake_ok) {
+		PRINT_DEBUG_MSG (1, "Waiting for connection from client, socket fd=%s.\n", socket_fd);
+		sock_accept (atoi(socket_fd), 4, &conn_fd);
```

Comment:
The return of `sock_accept` is always valid, that is why we need to try to write.

Reply:
Is that documented somewhere? I'm looking at https://docs.rs/wasi/latest/src/wasi/lib_generated.rs.html#2075-2082 but is that the correct code? That code suggests that it can return an error. I don't know what code would be surfacing this to an i32 return code.